Testplan WIP

GregAC opened the pull request:
~~Note the functional coverage plan referred to is still a PR: https://github.com/lowRISC/ibex/pull/1356 so the links are broken`~~

Functional coverage PR is merged to this has been updated to refer to it.

vogelpi commented on doc/03_reference/images/tb2.svg:
```diff
@@ -0,0 +1,989 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
The figure is hardly readable in the html. If you have time, you can maybe reduce the left and right borders as well as the empty space on the right inside the violet box.

vogelpi commented on doc/03_reference/testplan.rst:
```diff
@@ -0,0 +1,109 @@
+  This provides a driver and a monitor for the :ref:`Ibex Memory Interface Protocol<lsu-protocol>`.
+  The driver provides fully randomised and configurable timings for responses and randomisation of error responses.
+  Two agents are instaniated, one for the data memory interface the other for the instruction memory interface.
+  Read data for memory responses is provided from a backing memory, writes requests update the contents of the backing memory.
```
Nit: writes requests -> write requests

rswarbrick replied:
Also, probably a semicolon between the two halves.

vogelpi commented on doc/03_reference/testplan.rst:
```diff
@@ -0,0 +1,109 @@
+* Instructions to execute, these are generated by the `RISC-V DV random instruction <https://github.com/google/riscv-dv>`_ generator and provided to the testbench via a raw binary file.
+* Activity on external interfaces.
+
+Instructions are generated ahead of the test run so the test has no control over them.
+All external interfaces have their stimulus generated during the test run so can be controlled by the test.
```
The combination of "test run" (the execution itself) and "test" (some form of test specification) may be confusing here. Also, the test actually specifies which instructions will be tested (e.g. bitmanip yes/no, wfi etc.), i.e., the test has an influence on the instructions, it's just offline.

Would it make sense to rephrase this as follows?
"Instructions are generated ahead of time so the test has no control over them at run time."
"All external interfaces have their stimulus generated at run time so can be controlled by the test."

vogelpi reviewed (approved):
LGTM @GregAC ! I have just minor comments, one on the differentiation between "test" and "test run". But I am no verification expert - if you decide to merge this as is it's completely fine with me.

rswarbrick commented on doc/03_reference/testplan.rst:
```diff
@@ -0,0 +1,109 @@
+Co-simulation Agent:
+  This has multiple monitors.
+  One monitors the RVFI interface which provides details of retired instructions.
+  The other monitors relate to fetched instructions and instruction memory errors, more details are provided in :ref:`cosim`.
```
Nit: Should be semicolon before "more"

rswarbrick commented on doc/03_reference/testplan.rst:
```diff
@@ -0,0 +1,109 @@
+Memory Interface Agent
+  This provides a driver and a monitor for the :ref:`Ibex Memory Interface Protocol<lsu-protocol>`.
+  The driver provides fully randomised and configurable timings for responses and randomisation of error responses.
+  Two agents are instaniated, one for the data memory interface the other for the instruction memory interface.
```
Nit: instantiated (and probably follow it with a colon, not a comma)

rswarbrick commented on doc/03_reference/testplan.rst:
```diff
@@ -0,0 +1,109 @@
+The testbench instantiates the agents described above along with the memory model used by both the data and instruction side memory agents.
+A test consists of executing a pre-built binary (which is loaded into the memory model at the start of the test via backdoor accesses) along with configuring agents to provide appropriate stimulus for the test.
+Some tests may use the agents to generate stimulus at particular times (e.g.  interrupts).
+A test may perform additional checking on top of the co-simulation golden model comparison where appropriate (e.g. ensuring a raised interrupt has cause an exception).
```
Nit: caused

rswarbrick commented on doc/03_reference/testplan.rst:
```diff
@@ -0,0 +1,109 @@
+
+Stimulus falls into two categories:
+
+* Instructions to execute, these are generated by the `RISC-V DV random instruction <https://github.com/google/riscv-dv>`_ generator and provided to the testbench via a raw binary file.
```
Nit: Probably a colon before "these"

rswarbrick commented on doc/03_reference/testplan.rst:
```diff
@@ -0,0 +1,109 @@
+Tests
+-----
+
+As with stimulus test sequence development uses a coverage based approach.
```
I think you want a comma after "stimulus"? (I genuinely struggled to understand this sentence!)

rswarbrick commented on doc/03_reference/testplan.rst:
```diff
@@ -0,0 +1,109 @@
+
+As with stimulus test sequence development uses a coverage based approach.
+Tests will be added such that all coverage in the :ref:`coverage-plan` can be hit.
+All the details of specific tests will not be documented here.
```
Nit: This should probably be "Not all the details ... will be documented here."

rswarbrick commented on doc/03_reference/testplan.rst:
Nit: "one o**r** more" ?

rswarbrick reviewed (approved):
Looks good, modulo a couple of grammar nits.

GregAC commented:
Thanks for the reviews @rswarbrick and @vogelpi I've addressed all your comments.

As part of making the testbench diagram more readable I've expanded the width of the documentation in general (from max of 800 to 1000), which is the other commit of the PR.

vogelpi commented:
Very nice @GregAC , it's a good idea to change to the max width.
